### PR TITLE
Polyhedron demo: make several plugins work with Surface_mesh

### PR DESCRIPTION
**Pull request by afabri**
Addresses Issue #1931.  This PR is work in progress.

**I will split it in smaller pull requests to make integration smoother**

The following plugins work with `Surface_mesh`
- [ ] Plugins/AABB_tree/Cut_plugin.cpp
- [X] Plugins/PCA/Create_bbox_mesh_plugin.cpp
- [X] Plugins/PCA/Pca_plugin.cpp
- [ ] Plugins/PCA/Scene_polyhedron_transform_item.h
- [X] Plugins/PCA/Edit_box_plugin.cpp
- [ ] Plugins/Mesh_3/Detect_sharp_edges_plugin.cpp
- [ ] Plugins/Mesh_3/Optimization_plugin_cgal_code.cpp
- [ ] Plugins/Mesh_3/Mesh_3_plugin.cpp
- [ ] Plugins/IO/OFF_io_plugin.cpp
- [ ] Plugins/IO/GOCAD_io_plugin.cpp
- [ ] Plugins/IO/Surf_io_plugin.cpp
- [ ] Plugins/IO/STL_io_plugin.cpp
- [ ] Plugins/IO/VTK_io_plugin.cpp
- [ ] Plugins/Convex_decomposition/Nef_plugin.cpp
- [ ] Plugins/Surface_mesh_deformation/Edit_polyhedron_plugin.cpp
- [ ] Plugins/Surface_mesh_deformation/Scene_edit_polyhedron_item.h
- [ ] Plugins/Point_set/Surface_reconstruction_plugin.cpp
- [X] Plugins/Point_set/Point_set_shape_detection_plugin.cpp
- [ ] Plugins/Surface_mesh/Remeshing_plugin.cpp
- [ ] Plugins/Surface_mesh/Mesh_simplification_plugin.cpp
- [X] Plugins/Surface_mesh/Mesh_segmentation_plugin.cpp  depends on PR #1954 
- [ ] Plugins/Surface_mesh/Shortest_path_plugin.cpp
- [ ] Plugins/Surface_mesh/Remeshing_plugin_cgal_code.cpp
- [ ] Plugins/Surface_mesh/Parameterization_plugin.cpp
- [X] Plugins/Convex_hull/Convex_hull_plugin.cpp
- [X] Plugins/Convex_hull/Kernel_plugin.cpp
- [ ] Plugins/Operations_on_polyhedra/Clip_polyhedron_plugin.cpp
- [ ] Plugins/Operations_on_polyhedra/Point_set_from_vertices_plugin.cpp
- [X] Plugins/PMP/Inside_out_plugin.cpp
- [X] Plugins/PMP/Point_inside_polyhedron_plugin.cpp
- [X] Plugins/PMP/Polyhedron_slicer_plugin.cpp
- [X] Plugins/PMP/Corefinement_plugin.cpp
- [X] Plugins/PMP/Polyhedron_stitching_plugin.cpp
- [X] Plugins/PMP/Repair_polyhedron_plugin.cpp
- [X] Plugins/PMP/Surface_intersection_plugin.cpp
- [X] Plugins/PMP/Distance_plugin.cpp
- [ ] Plugins/PMP/Selection_plugin.cpp
- [ ] Plugins/PMP/Join_and_split_polyhedra_plugin.cpp
- [X] Plugins/PMP/Hole_filling_polyline_plugin.cpp
- [ ] Plugins/PMP/Hole_filling_plugin.cpp
- [X] Plugins/PMP/Jet_fitting_plugin.cpp
- [ ] Plugins/PMP/Isotropic_remeshing_plugin.cpp
- [X] Plugins/PMP/Mean_curvature_flow_skeleton_plugin.cpp  depends on PR #1954 
- [X] Plugins/PMP/Triangulate_facets_plugin.cpp
- [X] Plugins/Mesh_2/Mesh_2_plugin.cpp
- [ ] Plugins/Subdivision_methods/Subdivision_methods_plugin.cpp

### CODE REVIEW

**Comment by afabri** [comment]
@sloriot @maxGimeno   I have rebased the branch, so you have to reset  your local version

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Mean_curvature_flow_skeleton_plugin.cpp [hunk 1]
```diff
@@ -28,19 +32,47 @@
...
+
+template<>
+void set_halfedgeds_items_id (Scene_face_graph_item::FaceGraph&)
+{}
+
+} // namespace CGAL
+
+template <>
+void
+set_item_is_multicolor(typename Scene_face_graph_item*, bool)
```
Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Mean_curvature_flow_skeleton_plugin.cpp]: No typename here

**Comment by maxGimeno** [comment]
PCA_plugin does not compile.

**Comment by maxGimeno** [comment]
When trying to export to Polyhedron_item in Edit_box_plugin, there is an assertion.

**Comment by maxGimeno** [comment]
Mesh_segmentation_plugin does not compile

**Comment by maxGimeno** [comment]
Kernel_plugin crashes with SEGFAULT.

**Comment by maxGimeno** [comment]
Jet_fitting_plugin does not compile

**Comment by maxGimeno** [comment]
Mean_curvature_flow_skeleton_sm_plugin does not compile

**Comment by maxGimeno** [comment]
Mesh_2 plugin crashes at runtime.

**Comment by maxGimeno** [comment]
I did not test hole filling as you marked it as not finished

**Comment by afabri** [comment]
>> Mesh_2 plugin crashes at runtime.
It works with my data sets.

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Mean_curvature_flow_skeleton_plugin.cpp (hunk 1 above)
Comment (afabri) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Mean_curvature_flow_skeleton_plugin.cpp]: please fix/commit/push

**Comment by afabri** [comment]
The kernel_plugin works for cross.off with Polyhedron and Surface_mesh.

**Comment by afabri** [comment]
>> When trying to export to Polyhedron_item in Edit_box_plugin, there is an assertion.
I only have a release mode version of the demo, so we have to look on your machine what happens.

**Comment by maxGimeno** [comment]
>The kernel_plugin works for cross.off with Polyhedron and Surface_mesh.

Indeed, but not with elephant.off. Maybe it never did, though.

**Comment by maxGimeno** [comment]
For assertions, you can rebase on the last version of https://github.com/CGAL/cgal/pull/1954.

**Comment by lrineau** [comment]
@afabri This PR should be rebased on the current version of #1954.

**Comment by afabri** [comment]
@maxGimeno can you please do the rebasing.

**Comment by maxGimeno** [comment]
travis failed

**Comment by sloriot** [comment]
I've checked the corefinement plugin and there is no longer the crash.

**Comment by maxGimeno** [comment]
This one should be ready for integration.

File: Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt [hunk 2]
```diff
@@ -90,10 +111,14 @@ target_link_libraries(isotropic_remeshing_plugin  scene_polyhedron_item scene_po
...
+
 polyhedron_demo_plugin(detect_sharp_edges_plugin Detect_sharp_edges_plugin)
 target_link_libraries(detect_sharp_edges_plugin scene_polyhedron_item)
 
 polyhedron_demo_plugin(detect_sharp_edges_sm_plugin Detect_sharp_edges_plugin)
 target_link_libraries(detect_sharp_edges_sm_plugin scene_surface_mesh_item)
 target_compile_definitions(detect_sharp_edges_sm_plugin PUBLIC "-DUSE_SURFACE_MESH" )
+polyhedron_demo_plugin(distance_plugin Distance_plugin)
+target_link_libraries(distance_plugin scene_polyhedron_item  scene_color_ramp)
+target_compile_definitions(distance_sm_plugin PUBLIC "-DUSE_SURFACE_MESH" )
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt]: This plugin `distance_plugin` is already defined a few lines above!

I remove this PR from `integration`...

Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt]: I thought I fixed this during the rebasing....

Comment (MaelRL) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt]: The PR is still in integration for `lc-48`. I didn't remove it, but I cherry-picked 8a84263.

**Comment by lrineau** [comment]
This PR conflicts with #1969: they share a commit. Let's way and rebase this PR #1961 once the other ones are merged.

**Comment by maxGimeno** [comment]
No, I'd say we should close #1969, as if I understand correctly it does the same thing as what is done here. I guess #1969  is exctracted from this PR.

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_2/Mesh_2_plugin.cpp [hunk 3]
```diff
@@ -340,15 +341,27 @@ class Polyhedron_demo_mesh_2_plugin :
...
-                       *poly_item->polyhedron(),
-                       constant_coordinate_index,
-                       constant_coordinate);
-    scene->addItem(poly_item);
-    poly_item->invalidateOpenGLBuffers();
-
+    
+    bool build_polyhedron = false;
+    if (build_polyhedron){
+      Scene_polyhedron_item* poly_item = new Scene_polyhedron_item();
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_2/Mesh_2_plugin.cpp]: There is a logic error here. The `else` of that `if` will never be executed!

File: Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt [hunk 4]
```diff
@@ -15,12 +15,17 @@ if(EIGEN3_FOUND)
     polyhedron_demo_plugin(hole_filling_plugin Hole_filling_plugin ${hole_fillingUI_FILES})
     target_link_libraries(hole_filling_plugin scene_polyhedron_item scene_polylines_item scene_polyhedron_selection_item)
 
+    qt5_wrap_ui( hole_fillingUI_FILES Hole_filling_widget.ui)
+    polyhedron_demo_plugin(hole_filling_sm_plugin Hole_filling_plugin ${hole_fillingUI_FILES})
+    target_link_libraries(hole_filling_sm_plugin scene_surface_mesh_item scene_polylines_item scene_polyhedron_selection_item)
+target_compile_definitions(hole_filling_sm_plugin PUBLIC "-DUSE_SURFACE_MESH" )
+
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt]: Indentation issue

File: Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt [hunk 5]
```diff
@@ -47,17 +52,29 @@ target_link_libraries(mean_curvature_flow_skeleton_plugin
...
+#qt5_wrap_ui( Mean_curvature_flow_skeleton_pluginUI_FILES  Mean_curvature_flow_skeleton_plugin.ui)
+#polyhedron_demo_plugin(mean_curvature_flow_skeleton_sm_plugin Mean_curvature_flow_skeleton_plugin ${Mean_curvature_flow_skeleton_pluginUI_FILES})
+#target_link_libraries(mean_curvature_flow_skeleton_sm_plugin
+#    scene_surface_mesh_item
+#    scene_points_with_normal_item
+#    scene_polylines_item
+#    demo_framework)
+#target_compile_definitions(mean_curvature_flow_skeleton_sm_plugin PUBLIC "-DUSE_SURFACE_MESH" )
+
+
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/CMakeLists.txt]: Maybe you should re-read that comment, and either:
  - remove it,
  - or uncomment it.

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_2/Mesh_2_plugin.cpp [hunk 6]
```diff
@@ -340,15 +341,26 @@ class Polyhedron_demo_mesh_2_plugin :
...
-    poly_item->setName(iname);
-    cdt2_to_face_graph(cdt,
-                       *poly_item->polyhedron(),
-                       constant_coordinate_index,
-                       constant_coordinate);
-    scene->addItem(poly_item);
-    poly_item->invalidateOpenGLBuffers();
-
+    
+    if(mw->property("is_polyhedron_mode").toBool()){{
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_2/Mesh_2_plugin.cpp]: The double `{{` is an error. The error was [caught by Travis][travis].

**Comment by lrineau** [comment]
There is a link error on Windows:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-60/Polyhedron_Demo/TestReport_jtournois_x64_Cygwin-Windows10_MSVC2015-Release-64bits.gz

Note also that the error appeared while the testsuite was compiling the `all` target. That means that the file `Polyhedron/demo/Polyhedron/cgal_test_with_cmake` does not declare all the new targets, like those with the `_sm_` infix.

**Comment by lrineau** [comment]
I just merged a dozen of PR in `master`, and now this one has a conflict.

**Comment by maxGimeno** [comment]
Rebased

**Comment by maxGimeno** [comment]
This PR is transfered to #2242.  I merged it in my branch to fix some conflicts and it allows me to get the "parenthood" on the PR, allowing me to see its evolution in Travis, for example. 

This PR can therefore be closed.

**Comment by lrineau** [comment]
Merged into #2242, and thus replaced.

**Pull request closed without merging**